src/bit-systems/scene-preview-camera-system.ts

```diff
@@ -0,0 +1,119 @@
+import { getStreamerCamera } from "../components/scene-preview-camera";
+import { lerp } from "three/src/math/MathUtils";
+
+const startPositions = new Map<number, Vector3>();
+const startRotations = new Map<number, Quaternion>();
+const targetPositions = new Map<number, Vector3>();
+const targetRotations = new Map<number, Quaternion>();
+const startTimes = new Map<number, number>();
+const backwards = new Map<number, boolean>();
+const ranOnePass = new Map<number, boolean>();
+const newRot = new Quaternion();
+
+const updatePreviewCamera = (eid: number, previewCamera: Camera, cameraSystem: CameraSystem) => {
+  cameraSystem.mode = CAMERA_MODE_SCENE_PREVIEW;
+  const streamerCamera = getStreamerCamera();
+  if (streamerCamera) {
+    setMatrixWorld(previewCamera, streamerCamera.matrixWorld);
+    // Move camera forward just a bit so that we don't see the avatar's eye cylinders.
+    previewCamera.translateZ(-0.1);
+    previewCamera.matrixNeedsUpdate = true;
+  } else {
+    const startTime = startTimes.get(eid)!;
+    const duration = ScenePreviewCamera.duration[eid];
+    let t = (performance.now() - startTime) / (1000.0 * duration);
+    t = Math.min(1.0, Math.max(0.0, t));
+
+    if (!ranOnePass.get(eid)) {
+      t = t * (2 - t);
+    } else {
+      t = t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;
+    }
+
+    const startPosition = startPositions.get(eid)!;
+    const targetPosition = targetPositions.get(eid)!;
+    const startRotation = startRotations.get(eid)!;
+    const targetRotation = targetRotations.get(eid)!;
+
+    const back = backwards.get(eid);
+    const from = back ? targetPosition : startPosition;
+    const to = back ? startPosition : targetPosition;
+    const fromRot = back ? targetRotation : startRotation;
+    const toRot = back ? startRotation : targetRotation;
+
+    newRot.slerpQuaternions(fromRot, toRot, t);
+
+    previewCamera.position.set(lerp(from.x, to.x, t), lerp(from.y, to.y, t), lerp(from.z, to.z, t));
+
+    if (!ScenePreviewCamera.positionOnly[eid]) {
+      previewCamera.rotation.setFromQuaternion(newRot);
+    }
+
+    previewCamera.matrixNeedsUpdate = true;
+
+    if (t >= 0.9999) {
+      ranOnePass.set(eid, true);
+      backwards.set(eid, !back);
+      startTimes.set(eid, performance.now());
+    }
+  }
+};
+
+const scenePreviewCameraQuery = defineQuery([ScenePreviewCamera]);
+const scenePreviewCameraEnterQuery = enterQuery(scenePreviewCameraQuery);
+const scenePreviewCameraExitQuery = exitQuery(scenePreviewCameraQuery);
+export function scenePreviewCameraSystem(world: HubsWorld, cameraSystem: CameraSystem) {
+  scenePreviewCameraExitQuery(world).forEach(eid => {
+    startPositions.delete(eid);
+    startRotations.delete(eid);
+    targetPositions.delete(eid);
+    targetRotations.delete(eid);
+    startTimes.delete(eid);
+    backwards.delete(eid);
+    ranOnePass.delete(eid);
+  });
+  scenePreviewCameraEnterQuery(world).forEach(eid => {
+    const previewCamera = world.eid2obj.get(eid)!;
```

Comment:
This comment doesn't block the PR but is for future clean up.

If a system accesses Object3D, adding Object3DTag component to query would be safer (and less confusing for new readers)?

```typescript
const scenePreviewCameraQuery = defineQuery([ScenePreviewCamera, Object3DTag]);
```